### PR TITLE
[LINALG] Add E2E support for `aten.mean.dim` op

### PR DESCRIPTION
**Pull request by Shukla-Gaurav**
- This commit adds support for `aten.mean.dim` op.
- It also adds a new test script `stats.py` for statistics related ops.

Signed-Off-by: Gaurav Shukla <gaurav@nod-labs.com>

### CODE REVIEW

**Comment by Shukla-Gaurav** [comment]
@cathyzhyi @ramiro050 ready for re-review. thanks :)

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp [hunk 1]
```diff
@@ -34,11 +34,62 @@ static int getTensorRank(Value tensor) {
   return tensorRank;
 }
 
+// Helper function to compute the return type of the reduction function along
+// multiple dimensions. `dimList` specifies a list of dimensions to reduce and
+// `keepDim` preserves the rank of the input tensor.
+static LogicalResult
```
Comment (ramiro050) [inline lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:40]: nit: You can use the return type `FailureOr<BaseTensorType>` to return a `rewriter.notifyMatchFailure` or a `BaseTensorType`, so that you don't have to pass the return value as an argument to the function. You can then check if the returned value failed by doing `if (failed(LogicalResult(returnedValue)))`

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp [hunk 2]
```diff
@@ -34,11 +34,62 @@ static int getTensorRank(Value tensor) {
...
+          .cast<BaseTensorType>();
+  return success();
+}
+
 // Helper function to compute the return type of the reduction function.
 // `dim` specifies the dimension to reduce and `keepDim` preserves the rank of
 // the input tensor.
-static Type computeReductionType(PatternRewriter &rewriter, Operation *op,
-                                 Value input, Value dim, bool keepDim) {
+static Type computeReductionAlongDimType(PatternRewriter &rewriter,
```
Comment (ramiro050) [inline lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:90]: This is just a special case of the `computeReductionAlongMultipleDimsType`, so you should rewrite it using that function.

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp [hunk 3]
```diff
@@ -741,6 +797,131 @@ class DecomposeAtenMeanOp : public OpRewritePattern<AtenMeanOp> {
...
+
+    // Compute sum along dimensions specified in `dimList`.
+    Value none = rewriter.create<Torch::ConstantNoneOp>(loc);
+    Value sumAlongDims = rewriter.create<AtenSumDimIntListOp>(
+        loc, outputType, input, dimList, keepDim, /*dtype=*/none);
+
+    auto dimListConstruct = dimList.getDefiningOp<PrimListConstructOp>();
+    if (!dimListConstruct)
+      return rewriter.notifyMatchFailure(
+          op, "expect dimList to be constructed from list construct");
```
Comment (ramiro050) [inline lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:828]: You should check this before creating any ops to avoid doing extra work. Also, the `if` needs curly braces

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp [hunk 4]
```diff
@@ -741,6 +797,131 @@ class DecomposeAtenMeanOp : public OpRewritePattern<AtenMeanOp> {
...
+    BaseTensorType inputType = input.getType().cast<BaseTensorType>();
+    if (!inputType.hasDtype() || !inputType.getDtype().isa<mlir::FloatType>()) {
+      return rewriter.notifyMatchFailure(
+          op, "only floating-point type is supported");
+    }
+
+    // Compute sum along dimensions specified in `dimList`.
+    Value none = rewriter.create<Torch::ConstantNoneOp>(loc);
+    Value sumAlongDims = rewriter.create<AtenSumDimIntListOp>(
+        loc, outputType, input, dimList, keepDim, /*dtype=*/none);
```
Comment (ramiro050) [inline lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:823]: `AtenMeanDimOp` has a `dtype` argument that should be passed along into `AtenSumDimIntListOp`

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp [hunk 5]
```diff
@@ -741,6 +797,131 @@ class DecomposeAtenMeanOp : public OpRewritePattern<AtenMeanOp> {
...
+  LogicalResult matchAndRewrite(AtenMeanDimOp op,
+                                PatternRewriter &rewriter) const override {
+    Location loc = op.getLoc();
+    Value input = op.self();
+    Value dimList = op.dim();
+    Value keepDim = op.keepdim();
+    Type outputType = op.getType();
+
+    BaseTensorType inputType = input.getType().cast<BaseTensorType>();
+    if (!inputType.hasDtype() || !inputType.getDtype().isa<mlir::FloatType>()) {
```
Comment (ramiro050) [inline lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:815]: I think you should also check that `op.dtype()` is float. I tried passing a type different from float, and PyTorch gives casting errors.

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp [hunk 6]
```diff
@@ -741,6 +797,131 @@ class DecomposeAtenMeanOp : public OpRewritePattern<AtenMeanOp> {
...
+      return rewriter.notifyMatchFailure(op, "expect unbiased to be constant");
+    }
+
+    // Compute square(input - mean(input, dimList, keepdim=True)). Here,
+    // `keepdim` is set to True to ensure that the substraction is broadcast
+    // compatible.
+    Value none = rewriter.create<Torch::ConstantNoneOp>(loc);
+    Value constantTrue = rewriter.create<Torch::ConstantBoolOp>(loc, true);
+    BaseTensorType reduceType;
+    if (failed(computeReductionAlongMultipleDimsType(
```
Comment (ramiro050) [inline lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:879]: `if` needs curly braces

**Review by ramiro050 — CHANGES_REQUESTED** [review]

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp [hunk 7]
```diff
@@ -839,6 +851,54 @@ class DecomposeAtenMeanOp : public OpRewritePattern<AtenMeanOp> {
...
+    Value keepDim = op.keepdim();
+    Value dtype = op.dtype();
+    Type outputType = op.getType();
+    MLIRContext *context = op.getContext();
+
+    BaseTensorType inputType = input.getType().cast<BaseTensorType>();
+    if (!inputType.hasDtype() || !inputType.getDtype().isa<mlir::FloatType>() ||
+        !isNoneOrFloatDtype(context, dtype)) {
+      return rewriter.notifyMatchFailure(
+          op, "only floating-point type is supported");
```
Comment (ashay) [inline lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:874]: Just out of curiosity, what's our rationale behind choosing between `notifyMatchFailure()` versus `emitError()`?  Should we use `emitError()` here, so that the end user does not need to re-run the compilation with the `--debug` flag?

**Comment by Shukla-Gaurav** [comment]
@cathyzhyi @ramiro050 Gentle reminder.

File: python/torch_mlir_e2e_test/test_suite/stats.py [hunk 8]
```diff
@@ -0,0 +1,253 @@
...
+    def __init__(self):
+        super().__init__()
+
+    @export
+    @annotate_args([
+        None,
+        ([3, 4], torch.float32, True),
+    ])
+    def forward(self, x):
+        return torch.mean(x)
```
Comment (cathyzhyi) [inline python/torch_mlir_e2e_test/test_suite/stats.py:24]: prefer to use `torch.ops.aten.` operators directly.

**Review by cathyzhyi — APPROVED** [review]
Looks mostly good except we would prefer to use the aten operator directly in e2e tests.

**Review by ramiro050 — APPROVED** [review]
LGTM!

File: python/torch_mlir_e2e_test/test_suite/stats.py (hunk 8 above)
Comment (Shukla-Gaurav) [reply python/torch_mlir_e2e_test/test_suite/stats.py]: Thanks, done :)
